src/estimagic/optimization/_trustregion_bounded_newton_quadratic.py

```diff
@@ -74,224 +93,259 @@ def take_preliminary_steepest_descent_step_and_check_for_solution(
+
+                step_inactive = minimize_trust_conjugate_gradient(
+                    gradient_inactive, hessian_inactive, trustregion_radius
+                )
+                step_norm = np.linalg.norm(step_inactive)
+
+                if step_norm == 0:
+                    raise ValueError("Initial direction is zero.")
+
+        conjugate_gradient_step[active_bounds_info.inactive] = np.copy(step_inactive)
+        conjugate_gradient_step = _apply_bounds_to_conjugate_gradient_step(
+            conjugate_gradient_step,
+            x_candidate,
+            lower_bounds,
+            upper_bounds,
+            active_bounds_info,
+        )
 
     return (
-        x_inactive_cg,
-        gradient_inactive_cg,
-        gradient_norm_cg,
         conjugate_gradient_step,
+        step_inactive,
+        step_norm,
     )
 
 
 def compute_predicted_reduction_from_conjugate_gradient_step(
-    x_inactive_cg,
-    gradient_inactive_cg,
     conjugate_gradient_step,
-    gradient_candidate,
+    conjugate_gradient_step_inactive,
+    gradient_unprojected,
+    gradient_inactive,
     hessian_inactive,
     active_bounds_info,
 ):
     """Compute predicted reduction induced by the Conjugate Gradient step."""
     if active_bounds_info.all.size > 0:
         # Projection changed the step, so we have to recompute the step
         # and the predicted reduction. Leave the rust radius unchanged.
-        x_inactive_recomp = conjugate_gradient_step[active_bounds_info.inactive]
-        gradient_inactive_recomp = gradient_candidate[active_bounds_info.inactive]
+        cg_step_recomp = conjugate_gradient_step[active_bounds_info.inactive]
+        gradient_inactive_recomp = gradient_unprojected[active_bounds_info.inactive]
 
         predicted_reduction = evaluate_model_criterion(
-            x_inactive_recomp, gradient_inactive_recomp, hessian_inactive
+            cg_step_recomp, gradient_inactive_recomp, hessian_inactive
         )
     else:
         # Step did not change, so we can just recover the
         # pre-computed prediction
         predicted_reduction = evaluate_model_criterion(
-            x_inactive_cg,
-            gradient_inactive_cg,
+            conjugate_gradient_step_inactive,
+            gradient_inactive,
             hessian_inactive,
         )
 
-        predicted_reduction = -predicted_reduction
+    return -predicted_reduction
 
-    return predicted_reduction
 
-
-def perform_steepest_descent_and_update_trustregion_radius(
+def perform_gradient_descent_and_update_trustregion_radius(
     x_candidate,
-    f_candidate,
+    f_candidate_initial,
     gradient_projected,
     hessian_inactive,
     model,
     lower_bounds,
     upper_bounds,
     active_bounds_info,
     maxiter_steepest_descent,
+    options_update_radius,
 ):
     """Perform steepest descent step and update trust-region radius."""
-    x_trial = np.copy(x_candidate)
-    f_min = np.copy(f_candidate)
+    f_min = copy(f_candidate_initial)
```

Comment:
You make many copies of arrays at places where it is not immediately clear why copying is needed. Are they all necessary or just defensive? If they are defensive, I would prefer to simply not do inplace changes anywhere and get rid of all the copying.

Also, you use `copy`, `np.copy` and `arr.copy`. I would prefer to use `arr.copy` everywhere.

Reply:
Good point! Is a copy needed at all if we define a variable like this 

```python
if converged:
   hessian_inactive = model.square_terms

else:
   hessian_inactive = find_hessian_submatrix_where_bounds_inactive(
      model, active_bounds_info
      )
```
where `model` is a `namedtuple`?

Reply:
It depends. In the `if` clause you would need a copy if later there is any place where `hessian_inactive` gets changed in place but you do not want to have that change in `model.square_terms`, e.g. if you set some elements of `hessian_inactive` to zero without making a copy first. 

That is why it is very hard to tell whether copies are needed just from looking at the definition, unless you embrace functional style where 1. everything happens in functions and 2. no function modifies its inputs. 

Below I make 3 examples and discuss why I prefer the last:

1. Really wrong: 
```python
hessian_inactive = model.square_terms 
# a lot of code here
hessian_inactive[0, 0] = 0
```
-> model.square_terms[0, 0] is zero

2. Correct but hard to maintain
```python
hessian_inactive = model.square_terms.copy()
# a lot of code here
hessian_inactive[0, 0] = 0
```
-> model.square_terms is not changed but there will come the moment where I want to make the code faster and wonder if I can remove the `.copy`. Then I have to read a lot of code to find that out.

3. Functional style
```
def set_first_element_to_zero(arr):
    arr = arr.copy()
    arr[0, 0] = 0
    return arr

hessian_inactive = model.square_terms
# a lot of code here
hessian_inactive = set_first_element_to_zero(hessian_inactive)
```
The reason why the last is best is that you directly see why the copy is needed.

In fact, unless you have an iterative algorithm that changes hessian_inactive multiple times it would be better to not even initialize it with `model.square_terms` and directly do `hessian_inactive = set_first_element_to_zero(model.square_terms).

Reply:
Thanks for the detailed explanation!